lib/lib-utils/src/main/java/org/sagebionetworks/util/PaginationIterator.java

```diff
@@ -26,16 +27,25 @@ public PaginationIterator(PaginationProvider<T> provider, long limit) {
 		this.limit = limit;
 		this.provider = provider;
 		offset = 0L;
+		isDone = false;
 	}
 	
 	@Override
 	public boolean hasNext() {
+		if(isDone) {
+			return false;
+		}
 		if(currentPage == null || !currentPage.hasNext()) {
 			// Fetch the next page
 			List<T> page = provider.getNextPage(limit, offset);
 			offset += limit;
 			currentPage = page.iterator();
-			return currentPage.hasNext();
+			boolean hasNext =  currentPage.hasNext();
```

Comment:
If you check the page.size() < limit you avoid querying once more when there are results and you reached the last page